compiler bug: invalid use of void expression

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V doctor:**
```
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 2700 Eight-Core Processor  
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version

getwd: C:\Users\Bay\Desktop\t
vmodules: C:\Users\Bay\.vmodules
vroot: C:\path\v
vexe: C:\path\v\v.exe
vexe mtime: 2022-04-05 08:34:19
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 509367b.829fed4

Git version: git version 2.33.0.windows.2
Git vroot status: weekly.2022.13-60-g829fed4a
.git/config present: true
thirdparty/tcc: N/A
``` 


<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**

`v run .`

```v
struct App {}

fn (a App) test_func() {}

fn main() {
	mut cmd := Command{
		name: 'cli'
		version: '0.0.1'
	}

	app := App{}

	mut add := Command{
		name: 'add'
		description: 'Add something.'
		execute: app.test_func()
	}

	cmd.add_command(add)
	cmd.setup()
	cmd.parse(os.args)
}
```

**What did you expect to see?**
An error of some kind.
 
**What did you see instead?**
```
C:\Users\Bay\AppData\Local\Temp\v_0\t.17121622841115694210.tmp.c: In function 'main__main':
C:\Users\Bay\AppData\Local\Temp\v_0\t.17121622841115694210.tmp.c:20034:284: error: invalid use of void expression
20034 |         cli__Command add = (cli__Command){.name = _SLIT("add"),.usage = (string){.str=(byteptr)"", .is_lit=1},.description = _SLIT("Add something."),.man_description = (string){.str=(byteptr)"", .is_lit=1},.version = (string){.str=(byteptr)"", .is_lit=1},.pre_execute = 0,.execute = (voidptr)main__App_test_func(app),.post_execute = 0,.disable_help = 0,.disable_man = 0,.disable_version = 0,.disable_flags = 0,.sort_flags = 0,.sort_commands = 0,.parent = 0,.commands = __new_array(0, 0, sizeof(cli__Command)),.flags = __new_array(0, 0, sizeof(cli__Flag)),.required_args = 0,.args = __new_array(0, 0, sizeof(string)),.posix_mode = 0,};
      |                                                                                                                                                                                                                                                                                            ^
...
==================

builder error:
==================
C error. This should never happen
```

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -377,6 +377,10 @@ pub fn (mut c Checker) struct_init(mut node ast.StructInit) ast.Type {
 				expected_type = field_info.typ
 				c.expected_type = expected_type
 				expr_type = c.expr(field.expr)
+				if expr_type == ast.void_type {
+					c.error('cannot assign `void` expression to field `$field_info.name`',
+						field.pos)
+				}
 				if !field_info.typ.has_flag(.optional) {
 					expr_type = c.check_expr_opt_call(field.expr, expr_type)
 				}

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -378,8 +378,7 @@ pub fn (mut c Checker) struct_init(mut node ast.StructInit) ast.Type {
 				c.expected_type = expected_type
 				expr_type = c.expr(field.expr)
 				if expr_type == ast.void_type {
-					c.error('cannot assign `void` expression to field `$field_info.name`',
-						field.pos)
+					c.error('`$field.expr` (no value) used as value', field.pos)
 				}
 				if !field_info.typ.has_flag(.optional) {
 					expr_type = c.check_expr_opt_call(field.expr, expr_type)

--- a/vlib/v/checker/tests/struct_field_init_with_void_expr_err.out
+++ b/vlib/v/checker/tests/struct_field_init_with_void_expr_err.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/struct_field_init_with_void_expr_err.vv:19:3: error: cannot assign `void` expression to field `execute`
+vlib/v/checker/tests/struct_field_init_with_void_expr_err.vv:19:3: error: `app.test_func()` (no value) used as value
    17 |         name: 'add'
    18 |         description: 'Add something.'
    19 |         execute: app.test_func()

--- a/vlib/v/checker/tests/struct_field_init_with_void_expr_err.out
+++ b/vlib/v/checker/tests/struct_field_init_with_void_expr_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/struct_field_init_with_void_expr_err.vv:19:3: error: cannot assign `void` expression to field `execute`
+   17 |         name: 'add'
+   18 |         description: 'Add something.'
+   19 |         execute: app.test_func()
+      |         ~~~~~~~~~~~~~~~~~~~~~~~~
+   20 |     }
+   21 |

--- a/vlib/v/checker/tests/struct_field_init_with_void_expr_err.vv
+++ b/vlib/v/checker/tests/struct_field_init_with_void_expr_err.vv
@@ -1,0 +1,25 @@
+import cli { Command }
+import os
+
+struct App {}
+
+fn (a App) test_func() {}
+
+fn main() {
+	mut cmd := Command{
+		name: 'cli'
+		version: '0.0.1'
+	}
+
+	app := App{}
+
+	mut add := Command{
+		name: 'add'
+		description: 'Add something.'
+		execute: app.test_func()
+	}
+
+	cmd.add_command(add)
+	cmd.setup()
+	cmd.parse(os.args)
+}

